ingester/src/buffer_tree/partition/resolver/sort_key.rs

```diff
@@ -37,34 +37,63 @@ impl SortKeyResolver {
         Backoff::new(&self.backoff_config)
             .retry_all_errors("fetch partition sort key", || async {
                 let mut repos = self.catalog.repositories().await;
+
+                // fetch partition
                 let partition = repos
                     .partitions()
                     .create_or_get(self.partition_key.clone(), self.table_id)
                     .await?;
 
-                let (sort_key, sort_key_ids) =
+                let (p_sort_key, p_sort_key_ids) =
                     (partition.sort_key(), partition.sort_key_ids_none_if_empty());
 
-                assert_eq!(
-                    sort_key.as_ref().map(|v| v.len()),
-                    sort_key_ids.as_ref().map(|v| v.len())
-                );
+                // fetch partition's table columns
+                let columns = repos.columns().list_by_table_id(self.table_id).await?;
+
+                // build sort_key from sort_key_ids and columns
+                let sort_key =
+                    build_sort_key_from_sort_key_ids_and_columns(&p_sort_key_ids, &columns);
+
+                // This is here to catch bugs and will be removed once the sort_key is removed from the partition
+                assert_eq!(sort_key, p_sort_key);
 
-                Result::<_, iox_catalog::interface::Error>::Ok((sort_key, sort_key_ids))
+                Result::<_, iox_catalog::interface::Error>::Ok((sort_key, p_sort_key_ids))
             })
             .await
             .expect("retry forever")
     }
 }
 
+// build sort_key from sort_key_ids and columns
+// panic if the sort_key_ids are not found in the columns
+pub(crate) fn build_sort_key_from_sort_key_ids_and_columns(
+    sort_key_ids: &Option<SortedColumnSet>,
+    columns: &[Column],
+) -> Option<SortKey> {
+    sort_key_ids.as_ref().map(|ids| {
+        let names = ids
+            .iter()
+            .map(|id| {
+                columns
+                    .iter()
+                    .find(|c| c.id == *id)
+                    .map(|c| c.name.clone())
+                    .expect("Cannot find column names for sort key ids")
+            })
+            .collect::<Vec<_>>();
+
+        SortKey::from_columns(names.iter().map(|s| &**s))
+    })
+}
```

Comment:
This could probably be written to consume a column iterator and simplify the iterator chaining a little. Haven't measured this change performance-wise (or looked in Godbolt), but thought I'd leave it here for discussion.

```suggestion
// build sort_key from sort_key_ids and columns
// panic if the sort_key_ids are not found in the columns
pub(crate) fn build_sort_key_from_sort_key_ids_and_columns<I>(
    sort_key_ids: &Option<SortedColumnSet>,
    columns: I,
) -> Option<SortKey>
where
    I: Iterator<Item = Column>,
{
    let mut column_names = columns.map(|c| (c.id, c.name)).collect::<hashbrown::HashMap<_, _>>();
    sort_key_ids.as_ref().map(|ids| {
        SortKey::from_columns(ids.iter().map(|id| {
            column_names
                .remove(id)
                .expect(format!("cannot find column names for sort key id {}", id.get()).as_str())
        }))
    })
}
```

Reply:
I think regardless it'd be good to include the missing ID in the panic message 👍

Reply:
Thanks @savage-engineer . I will use it

Reply:
Done